feat(internals): Add documentation for profiler usage (#921)
inveniosoftware/docs-invenio-rdm: docs/maintenance/internals/profiler.md

# Profiler

> Invenio App RDM integrates [Flask-MultiProfiler](https://github.com/inveniosoftware/Flask-MultiProfiler) for analyzing Python, SQL, and search performance.

Reviewer: ```suggestion
# Profiler

_Introduced in v14_

InvenioRDM integrates [Flask-MultiProfiler](https://github.com/inveniosoftware/Flask-MultiProfiler) for analyzing Python, SQL, and search performance.
```

## Features

- Start/stop profiling sessions from a web UI
- **Code profiling**: function-level profiling using [pyinstrument](https://pyinstrument.readthedocs.io/)
- **SQL profiling**: database query profiling using [sqltap](https://github.com/inconshreveable/sqltap)

> - **Search profiling**: OpenSearch/Elasticsearch operation tracing via client trace logging

Reviewer: ```suggestion
- **Search profiling**: OpenSearch operation tracing via client trace logging
```

- `MULTIPROFILER_PERMISSION`: callable returning `True` if the current user may access the profiler (default: administrators only via `administration_permission.can`)
- `MULTIPROFILER_SEARCH_TRACE_LOGGER`: logger name used for search trace logging (default: `opensearchpy.trace`)
- `MULTIPROFILER_BASE_TEMPLATE`: base template for the profiler page (default: `flask_multiprofiler/index.html`)

## Usage

With the profiler extra installed:

1. Sign in as a user allowed by `MULTIPROFILER_PERMISSION` (administrators by default).
2. Open the **Profiler** entry from the user profile admin menu, or visit `/profiler/` directly (e.g. `http://localhost:5000/profiler/`).
3. Start a profiling session by choosing:
    - a session ID,
    - whether to enable code profiling,
    - whether to enable SQL profiling,
    - whether to enable search profiling.
4. Interact with your site as usual. The profiler collects data for each request while the session is active.
5. Return to `/profiler/` to view and manage collected sessions and their reports.

![Profiler UI screenshot](./img/profiler.png)

![SQL profiling screenshot](./img/sql-profiling.png)

## Search profiling

> Search profiling intercepts trace logs from the OpenSearch or Elasticsearch Python client. InvenioRDM defaults to `opensearchpy.trace`.

Reviewer: ```suggestion
Search profiling intercepts trace logs from the OpenSearch Python client. InvenioRDM defaults to `opensearchpy.trace`.
```

Supported clients:

- OpenSearch
- Elasticsearch v7.x

> 

Reviewer: ```suggestion
```
